### PR TITLE
Build necessary extensions for release

### PR DESCRIPTION
**Pull request by baotingfang**
Add `--enable-debug-tools` configure option

  * when providing the --enable-debug-tools option, it will includes
    the following extensions:

      gp_distribution_policy
      gp_internal_tools
      gp_sparse_vector
      gp_replica_check
      gp_inject_fault
      gp_debug_numsegments

    otherwise, it will only inlude:

      gp_distribution_policy
      gp_internal_tools
      gp_sparse_vector

Co-authored-by: Bob Bao <bbao@pivotal.io>
Co-authored-by: Ning Fu <nfu@pivotal.io>

## Here are some reminders before you submit the pull request
- [ ] Add tests for the change
- [ ] Document changes
- [ ] Communicate in the mailing list if needed
- [ ] Pass `make installcheck`
- [ ] Review a PR in return to support the community

### CODE REVIEW

**Comment by kmacoskey** [comment]
I'm not sure if using an environment variable to drive the flow within the recursive make for gpcontrib modules is a good pattern or not. 

Have we considered some of these options:
1. Create a new configure flag, such as `--enable-debug-tools` (only an example), that will determine whether the gpcontrib modules for debug/dev/test are built.
2. Create new configure flags for each gpcontrib module we want to enable or disable, then making decisions about what is default and what is not, such as:
  * `--enable-sparse-vector`
  * `--enable-internal-tools`
  * `--enable-distribution-policy`

File: gpcontrib/Makefile [hunk 1]
```diff
@@ -20,7 +20,15 @@ recurse_targets = gp_sparse_vector \
...
                gp_internal_tools \
-               gp_inject_fault
+               gp_inject_fault \
+               gp_replica_check
+
+ifeq "${RELEASE}" "yes"
+	recurse_targets = gp_sparse_vector \
+                   gp_distribution_policy \
+                   gp_internal_tools
+endif
```
Comment (benchristel) [inline gpcontrib/Makefile]: To make it clearer that there are two mutually exclusive branches here, we should move lines 19-24 into an `else`.

```makefile
ifeq "${RELEASE}" "yes"
	recurse_targets = gp_sparse_vector \
                   gp_distribution_policy \
                   gp_internal_tools
else
    recurse_targets = gp_sparse_vector \
                   gp_debug_numsegments \
                   gp_distribution_policy \
                   gp_internal_tools \
                   gp_inject_fault
                   gp_inject_fault \
                   gp_replica_check
endif
```

**Comment by baotingfang** [comment]
> 1. Create a new configure flag, such as `--enable-debug-tools` (only an example), that will determine whether the gpcontrib modules for debug/dev/test are built.

@kmacoskey We think this solution is a better choice

File: gpcontrib/Makefile [hunk 2]
```diff
@@ -17,10 +17,14 @@ include $(top_builddir)/src/Makefile.global
 # targets built via top_builddir/GNUmakefile
 #
 recurse_targets = gp_sparse_vector \
-               gp_debug_numsegments \
                gp_distribution_policy \
-               gp_internal_tools \
-               gp_inject_fault
+               gp_internal_tools
```
Comment (benchristel) [inline gpcontrib/Makefile]: Why was `gp_replica_check` not in this list before?

Comment (kmacoskey) [reply gpcontrib/Makefile]: In which list exactly?

**Comment by benchristel** [comment]
The code looks good, though the tests are failing because the `gp_inject_fault` UDF doesn't exist. It looks like we need to build GPDB with `--enable-debug-tools` in the PR pipeline, and possibly other pipelines.

File: gpcontrib/Makefile (hunk 2 above)
Comment (benchristel) [reply gpcontrib/Makefile]: The `recurse_targets` list.

**Comment by baotingfang** [comment]
> The code looks good, though the tests are failing because the `gp_inject_fault` UDF doesn't exist. It looks like we need to build GPDB with `--enable-debug-tools` in the PR pipeline, and possibly other pipelines. @baotingfang @funcode

We've updated the PR pipeline to include the option --enable-debug-tools.

File: gpcontrib/Makefile (hunk 1 above)
Comment (soumyadeep2007) [reply gpcontrib/Makefile]: +1

File: gpAux/Makefile [hunk 3]
```diff
@@ -162,8 +162,8 @@ endif
 aix7_ppc_64_CONFIGFLAGS=--disable-gpcloud --without-readline --without-libcurl --disable-orca --disable-pxf --without-zstd $(APR_CONFIG)
 win32_CONFIGFLAGS=--with-gssapi --without-libcurl --disable-orca --disable-pxf --disable-gpcloud --without-libbz2 $(APR_CONFIG)
 sol10_x86_64_CONFIGFLAGS= --with-libxml $(APR_CONFIG)
-rhel6_x86_64_CONFIGFLAGS=--with-quicklz --enable-tap-tests --enable-gpperfmon --with-gssapi --enable-mapreduce --enable-orafce ${ORCA_CONFIG} --with-libxml $(APU_CONFIG)
-rhel7_x86_64_CONFIGFLAGS=--with-quicklz --enable-tap-tests --enable-gpperfmon --with-gssapi --enable-mapreduce --enable-orafce ${ORCA_CONFIG} --with-libxml $(APU_CONFIG)
+rhel6_x86_64_CONFIGFLAGS=--enable-debug-extensions --with-quicklz --enable-tap-tests --enable-gpperfmon --with-gssapi --enable-mapreduce --enable-orafce ${ORCA_CONFIG} --with-libxml $(APU_CONFIG)
+rhel7_x86_64_CONFIGFLAGS=--enable-debug-extensions --with-quicklz --enable-tap-tests --enable-gpperfmon --with-gssapi --enable-mapreduce --enable-orafce ${ORCA_CONFIG} --with-libxml $(APU_CONFIG)
```
Comment (benchristel) [inline gpAux/Makefile]: Is this the change that fixes the PR pipeline? How do we know it's not going to affect other pipelines and released versions of GPDB?

Comment (kmacoskey) [reply gpAux/Makefile]: Adding `--enable-debug-extensions` here would enable it in both this pipeline and the release pipeline which is undesirable.

I believe where you may want to include this flag is in these locations only:
https://github.com/greenplum-db/gpdb/blob/master/concourse/pipelines/gpdb_master-generated.yml#L739
https://github.com/greenplum-db/gpdb/blob/master/concourse/pipelines/gpdb_master-generated.yml#L781
https://github.com/greenplum-db/gpdb/blob/master/concourse/pipelines/gpdb_master-generated.yml#L814

`{{configure_flags}}` is an interpolated variable so you would need to add the flag where in the repository where that variable is defined.

Comment (baotingfang) [reply gpAux/Makefile]: We add a new variable to the pipeline: `configure_flags_with_extensions`. It should be set to  "--enable-cassert --enable-debug-extensions" and only for these tasks:

```
compile_gpdb_centos6
compile_gpdb_centos7
compile_gpdb_sles11
```

**Comment by danielgustafsson** [comment]
Please refrain from opening branches on the master repository, the intended workflow is to open all PRs from forks.

File: gpcontrib/Makefile [hunk 4]
```diff
@@ -16,11 +16,21 @@ include $(top_builddir)/src/Makefile.global
...
-               gp_debug_numsegments \
+recurse_targets = ""
+
+ifeq "$(enable_debug_extensions)" "yes"
+	recurse_targets = gp_sparse_vector \
                gp_distribution_policy \
                gp_internal_tools \
-               gp_inject_fault
+               gp_debug_numsegments \
+               gp_inject_fault \
```
Comment (danielgustafsson) [inline gpcontrib/Makefile]: I might be late to the party, but why would we want to ship `gp_inject_fault`  in a release?

**Comment by baotingfang** [comment]
> Please refrain from opening branches on the master repository, the intended workflow is to open all PRs from forks.

Sorry about this, That's a good practice, we will follow it for new PRs

File: gpcontrib/Makefile (hunk 4 above)
Comment (funcode) [reply gpcontrib/Makefile]: It is not for the product release, only for the dev release.

Comment (danielgustafsson) [reply gpcontrib/Makefile]: Gotcha, I got confused on the terminology. Thanks for clarifying.

**Comment by dyozie** [comment]
Is there a way we can ensure that the README file for `gp_sparse_vector` is included (copied over as part of the GPDB install)?  It would be great to be able to just reference it in place.

**Comment by bradfordboyle** [comment]
:shipit:

**Pull request merged**